### PR TITLE
TYP: misc return types

### PR DESCRIPTION
**Pull request by twoertwein**
xref #47521

Most changes are from renaming `T` to `_<class>T`.

### CODE REVIEW

**Review by MarcoGorelli — APPROVED** [review]
looks good to me

File: pandas/io/sql.py [hunk 1]
```diff
@@ -1364,7 +1364,7 @@ class PandasSQL(PandasObject, ABC):
     Subclasses Should define read_query and to_sql.
     """
 
-    def __enter__(self):
+    def __enter__(self) -> PandasSQL:
```
Comment (topper-123) [inline pandas/io/sql.py:1367]: This returns the relevant subclass (`SQLDatabase` & `SQLDatabase`) when those are used. Add a `_PandasSqlT = TypeVar("_PandasSqlT", bound="PandaSQL")` and use that?

I've made some arguments that we should allow use of `typing_extensions.Self` in some other threads, but that's a bit of a different issue.

Comment (twoertwein) [reply pandas/io/sql.py:1367]: I would love using Self here

**Comment by topper-123** [comment]
`typing_extensions` has just been merged into main in #51480. So you can update this  to use `from _typing import Self` instead of using `typevar`.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @twoertwein

**Pull request merged**